src/tqec/compile/substitute.py

```diff
@@ -0,0 +1,218 @@
+    during the block graph compilation.
+
+    Attributes:
+        spec1: the cube specification of the first cube. By convention, the cube corresponding
+            to `spec1` should have a smaller position than the cube corresponding to `spec2`.
+        spec2: the cube specification of the second cube.
+        pipe_type: the type of the pipe connecting the two cubes.
+    """
+
+    spec1: CubeSpec
+    spec2: CubeSpec
+    pipe_type: PipeType
+
+
+class SubstitutionRule(Protocol):
+    def __call__(
+        self,
+        key: SubstitutionKey,
+        block1: CompiledBlock,
+        block2: CompiledBlock,
+    ) -> tuple[CompiledBlock, CompiledBlock]:
+        """Substitute the two blocks with the given key.
+
+        Args:
+            key: `SubstitutionKey` that includes the necessary information for the substitution, including
+                the cube specifications and the pipe type that connecting the two blocks.
+            block1: the first block to substitute.
+            block2: the second block to substitute.
+        """
+        ...
+
+
+def default_substitution_rule(
+    key: SubstitutionKey,
+    block1: CompiledBlock,
+    block2: CompiledBlock,
+) -> tuple[CompiledBlock, CompiledBlock]:
+    """The default substitution rule for the block graph compilation."""
+    # 1. The spatial junctions are not supported yet
+    if key.spec1.is_spatial_junction or key.spec2.is_spatial_junction:
+        raise TQECException("Spatial junctions are not supported yet.")
+    # 2. Substitute the time-direction pipe
+    if key.pipe_type.direction == Direction3D.Z:
+        return _substitute_in_time_direction(key, block1, block2)
+    # 3. Substitute the space-direction pipe
+    else:
+        return _substitute_in_space_with_usual_cubes(key, block1, block2)
+
+
+DEFAULT_SUBSTITUTION_RULES: defaultdict[SubstitutionKey, SubstitutionRule] = (
+    defaultdict(lambda: default_substitution_rule)
+)
+
+
+def _substitute_in_time_direction(
+    key: SubstitutionKey,
+    block1: CompiledBlock,
+    block2: CompiledBlock,
+) -> tuple[CompiledBlock, CompiledBlock]:
+    assert key.pipe_type.direction == Direction3D.Z, "Pipe direction must be Z."
+    # Substitute the final layer of the bottom block with the bulk layer
+    bottom_block = block1.with_updated_layer(
+        plaquettes_to_update=deepcopy(block1.layers[1].collection),
+        layer_to_update=block1.num_layers - 1,
+    )
+    # Substitute the first layer of the top block with the bulk layer
+    top_block = block2.with_updated_layer(
+        plaquettes_to_update=deepcopy(block2.layers[1].collection),
+        layer_to_update=0,
+    )
+    # If the pipe has a hadamard, apply the hadamard transformation
+    # The hadamard transformation is applied to the end of the bottom block
+    if key.pipe_type.has_hadamard:
+        for plaquette in bottom_block.layers[-1].collection.values():
+            _inplace_append_timelike_hadamard_transition(plaquette)
```

Comment:
I am pretty sure that this Hadamard layer (that we will have to support eventually and that is correctly implemented here) breaks the assumptions made by the detector computation code that a `Fragment` is made of `(R -> [computation] ->) M` because the H gates, that are considered `[computation]` will appear before the resets of the next plaquette and no measurement will be present between the two.

Pushing the Hadamard gates inside the QEC rounds of the next block would be a solution, but that may lead to a shift in the `schedule` (not sure) and that might break some other assumptions on the code. 

This is something to check eventually.

Reply:
Oh I did not realize the problem. Pushing it into the QEC rounds seems doable because currently there is always a moment of H gates before measurement. But to do that, we may need to cancel out some double H on data qubits at that moment.

Reply:
I do think we can implement the hadamard transition(spatial/temporal) later after finalizing a better architecture. For now, it's fine to remove the relevant code of hadamard pipes. What do you think?

Reply:
I completely agree. I'll make a full comment on my thoughts about how we should proceed in my opinion.